### PR TITLE
Bug fix/get filedeck fromfile

### PR DESCRIPTION
**Pull request by WPringle**
`file_deck` and `advisory` properties should be found correctly when using the `from_file` command. 

This is affecting `EnsemblePerturbation` as it now requires to know the `file_deck` for doing the perturbations by forecast or hindcast.

### CODE REVIEW

**Comment by WPringle** [comment]
@SorooshMani-NOAA It is from `test_original_file` for the new PR:  [`EnsemblePerturbation`](https://github.com/noaa-ocs-modeling/EnsemblePerturbation/pull/93#issuecomment-1494690232).

**Comment by WPringle** [comment]
@SorooshMani-NOAA When you read a storm from file it was populating the `advisories` with just the list of possibilities and `file_deck` was empty. Not sure why the catch for` filename` being `None` was there.

**Comment by SorooshMani-NOAA** [comment]
I see, I think it might have been there to avoid assuming file type when a file is explicitly passed to the `VortexTrack` constructor; basically saying: "_you gave me the file, how would I know what deck it's from!_". I see merit in both having and not having that extra check. I think it might open a can of worms if we remove it, but I might just be too cautions!

If we modify `from_file` to get more arguments and set some stuff from the beginning, it might be safer. Because I see a bunch of checks for `self.advisories` which itself is kind of undefined if `from_file` is called.

@zacharyburnett do you have any suggestions?

**Comment by zacharyburnett** [comment]
> If we modify `from_file` to get more arguments and set some stuff from the beginning, it might be safer. Because I see a bunch of checks for `self.advisories` which itself is kind of undefined if `from_file` is called.

That makes sense to me, if `self.advisories` isn't populated then we should ask for more info / have asked for info in the parameters

**Comment by WPringle** [comment]
It should be able to determine the deck it is from because if its says `BEST` in the columns it is from file_deck `b`. Otherwise, `a`. I forget what the `f` is from. That's what it seems to be doing anyway.

**Comment by SorooshMani-NOAA** [comment]
In the constructor we have:
https://github.com/oceanmodeling/StormEvents/blob/123b9efa6244cda6ed2262476e99cb10245452ce/stormevents/nhc/track.py#L80-L81
and a bit later:
https://github.com/oceanmodeling/StormEvents/blob/123b9efa6244cda6ed2262476e99cb10245452ce/stormevents/nhc/track.py#L101-L102


My concern is that `from_file` initializes the `advisories` to `None`, which is set as (going to line 398):
https://github.com/oceanmodeling/StormEvents/blob/123b9efa6244cda6ed2262476e99cb10245452ce/stormevents/nhc/track.py#L395-L402

this checks for `__valid_advisories`, which itself depends on `file_deck`, that is again `None` at this point (when creating using `from_file`). So it goes to 407 below
https://github.com/oceanmodeling/StormEvents/blob/123b9efa6244cda6ed2262476e99cb10245452ce/stormevents/nhc/track.py#L405-L423

So you will have all the valid advisories in the list (including `BEST`). Now if we remove `self.filename` check it then proceeds to line 371 below
https://github.com/oceanmodeling/StormEvents/blob/123b9efa6244cda6ed2262476e99cb10245452ce/stormevents/nhc/track.py#L369-L380

so it always should say BEST in the end if we remove the check! At least that's how I understand from the code.

Even if it's **not** doing that, it means that the code is complicated enough to make me nervous for changing such condition.

**Comment by WPringle** [comment]
@SorooshMani-NOAA I agree it's complicated, I couldn't work it out what all those queries were doing. 
But straightforwardly, my opinion is that the `advisory` column in the `.data` is a property that can be queried to get the property as BEST or OFCL etc, then from that infer it is `file_deck` = `b` or `a`. Should be changed to do this simple operation.

**Comment by SorooshMani-NOAA** [comment]
@WPringle you're right about the logic for when we load from file. You can add a condition where `self.filename` is not `None`, then you can get `advisories` from the unique list of advisories from the loaded file (in the advisories setter method), similar for `file_deck` setter method. That way, the old conditions will stay the same, but we also cover cases where data is loaded from file.

**Comment by zacharyburnett** [comment]
there's always an edge case 😅

**Comment by SorooshMani-NOAA** [comment]
@WPringle this might be a buggy test then! Let's track it separately in a separate issue.

**Comment by WPringle** [comment]
@SorooshMani-NOAA @zacharyburnett I added the ability to specify the `file_deck` and `advisories` when read from file. By default, like when getting by storm code from NHC, it will be assumed to be best track. I still don't think the catch for `filename` is `None` is needed because it should just behave the same way as when getting from the internet.

**Comment by WPringle** [comment]
@SorooshMani-NOAA @zacharyburnett The only other thing I can think of is that we could add a warning that the `file_deck` or `advisories` entry is recommended when reading from file because the code doesn't know which it is.

File: stormevents/nhc/track.py [hunk 1]
```diff
@@ -146,11 +146,15 @@ def from_file(
         path: PathLike,
         start_date: datetime = None,
         end_date: datetime = None,
+        file_deck: ATCF_FileDeck = None,
+        advisories: [ATCF_Advisory] = None,
```
Comment (pmav99) [inline stormevents/nhc/track.py:150]: `[ATCF_Advisory]` is not a valid type declaration. 

If `advisories` is meant to be a list of `ATCF_Advisory` objects, then:
1. the docstring should be updated accordingly
2. the type declaration should be updated to:
``` python
from typing import List

        advisories: List[ATCF_Advisory = None,
```
In Python 3.9+ you can even omit the import and declare it as:
``` python
        advisories: list[ATCF_Advisory] = None,
```

That being said, if an explicit `list` is not required by the code, then `typing.Iterable` is probably a better choice.

If on the other hand, the docstring in line 157 is accurate then the declaration should be:
``` python
        advisories: ATCF_Advisory = None,
```

Comment (WPringle) [reply stormevents/nhc/track.py:151]: @SorooshMani-NOAA @zacharyburnett Which you think it is?

Comment (pmav99) [reply stormevents/nhc/track.py]: @WPringle I would also update the docstring. Maybe something like:

``` diff
- :param advisories: ATCF advisory type; one of ``BEST``, ``OFCL``, ``OFCP``, ``HMON``, ``CARQ``, ``HWRF``
+ :param advisories: list of ATCF advisory types; valid choices are: ``BEST``, ``OFCL``, ``OFCP``, ``HMON``, ``CARQ``, `HWRF``
 ```

**Comment by SorooshMani-NOAA** [comment]
@WPringle other than the docstring for the input list of advisories that @pmav99 mentioned (still not fixed) the implementation looks good. BTW, this repo is using `black` for styling (vs `oitnb` for `ensembleperturbation`)

**Comment by WPringle** [comment]
> @WPringle other than the docstring for the input list of advisories that @pmav99 mentioned (still not fixed) the implementation looks good. BTW, this repo is using `black` for styling (vs `oitnb` for `ensembleperturbation`)

how to do the `black` styling?
